tedana/utils.py

```diff
@@ -187,9 +190,17 @@ def make_adaptive_mask(data, mask, threshold=1, methods=["dropout"]):
 
         LGR.info("Echo-wise intensity thresholds for adaptive mask: %s", lthrs)
 
-        # determine samples where absolute value is greater than echo-specific thresholds
-        # and count # of echos that pass criterion
-        dropout_adaptive_mask = (np.abs(echo_means) > lthrs).sum(axis=-1)
+        # Find the last good echo for each voxel
+        dropout_adaptive_mask = np.zeros(n_samples, dtype=np.int16)
+        for i_voxel in range(n_samples):
+            echo_means_voxel = np.abs(echo_means[i_voxel, :])
+            echos_over_threshold = echo_means_voxel > lthrs
+            # Find the index of the last True element in a 1D boolean array
+            last_true_index = (
+                (np.where(echos_over_threshold)[0][-1] + 1) if np.any(echos_over_threshold) else 0
+            )
+            dropout_adaptive_mask[i_voxel] = last_true_index
+
```

Comment:
```suggestion
        # Find the last good echo for each voxel
        dropout_adaptive_mask = np.zeros(n_samples, dtype=np.int16)
        for echo_idx in range(n_echos):
            # For voxels that were in the mask for the immediately previous echo
            # If they are still good in the current echo, increment the adaptive
            # mask value
            dropout_adaptive_mask[
                (dropout_adaptive_mask == (echo_idx)) * (echo_means[:, echo_idx] > lthrs[echo_idx])
            ] = (echo_idx + 1)
        adaptive_masks.append(dropout_adaptive_mask)

```

Reply:
I added simulated data to the adaptive mask tests in the last couple commits, and your change doesn't produce results that I'd expect.

I have one simulated voxel that is [mean of first echo, 1, mean of first echo], which _should_ give a decay-based adaptive mask value of 3 (i.e., the mean value of the first echo should definitely be good in the third echo), but your change gives a value of 1.

Reply:
I think your test case is performing as I understand it should. Each echo has its own threshold. When the test is run `lthrs == [ 220.73333333, 79.73333333, 53.4  ]`
For your added test, the failing voxel is your `idx+1` where `data[idx+1,:,0] == [2442.976348096348, 1.0, 2442.976348096348]` That means the first and third echos are above threshold (`True`) but the second echo is below threshold (`False`). Since the second echo is bad, the last good echo is the first echo and the adaptive mask correctly equals 1.

Reply:
The last good echo in that case is the third echo, since 2442.976348096348 > 53.4. The descriptions from https://github.com/ME-ICA/tedana/issues/679#issuecomment-782423984 and https://github.com/ME-ICA/tedana/issues/679#issuecomment-1956766801 indicate that the last good echo should be used, rather than the last contiguous good echo.

Reply:
OK. Then this is a confusion on the specifications. I was thinking the "last good echo" was the one before the first bad echo while your earlier comments are saying to keep intermediate echoes if a later one is good under the assumption the threshold got something wrong in-between. That is:
```
good, good, good == 3
good, good, bad == 2
good, bad, good == (2 in current Main, 1 with my specification and 3 with your specification)
```
I think we agree that the current Main is wrong, I think either of our ideas is reasonable. I might want to dig a bit into real data to see when `good,bad,good` happens to see if the intermediate `bad` tends to be a sign that something is wrong with later echoes or if there are just cases where a middle echo is just below a threshold and we should bias towards using.

We also added my approach earlier for removing 0, negatives or Nan values. I think in that case, it makes sense to say, if we have `good, Nan, good` or `good, 0, good` the adapative mask should be 1 since the middle echo is unusable (Since we were talking past each other I wanted to make sure we both agree on that one)

https://github.com/ME-ICA/tedana/blob/682f8a61653af6a772d3f94f0b8c80f53c44bf93/tedana/utils.py#L151-L161

Reply:
> I think we agree that the current Main is wrong, I think either of our ideas is reasonable.

I agree that the sum is not the way to go, and I can see why using the last contiguous good echo makes the most sense (that was originally how I interpreted the adaptive mask method description), but I remember testing that approach and I ended up losing a lot of voxels. TBF, that might not be as much of an issue now that we have different for denoising and for ICA, but it's worth taking into account when we decide between the two proposed approaches here.

> We also added my approach earlier for removing 0, negatives or Nan values. I think in that case, it makes sense to say, if we have good, Nan, good or good, 0, good the adapative mask should be 1 since the middle echo is unusable

I also agree that having a zero or NaN in a later echo should limit the adaptive mask value. That's how it will currently work, since we take the voxel-wise minimum value from the different requested adaptive masks, and the "base" one will have a value of 1 here.

Reply:
I updated my above suggested code change to something that should match your specifications, but is a bit simpler to reach and avoids a for loop across voxels.